Quantum-Control-Applications/Superconducting/Single-Fixed-Transmon/04b_resonator_spectroscopy_wide_range_octave.py

```diff
@@ -0,0 +1,175 @@
+This frequency can be used to update the readout intermediate frequency in the configuration under "resonator_IF".
+
+Prerequisites:
+    - Ensure calibration of the time of flight, offsets, and gains (referenced as "time_of_flight").
+    - Calibrate the IQ mixer connected to the readout line (whether it's an external mixer or an Octave port).
+    - Define the readout pulse amplitude and duration in the configuration.
+    - Specify the expected resonator depletion time in the configuration.
+
+Before proceeding to the next node:
+    - Update the readout frequencies, labeled as "resonator_IF" and "resonator_LO", in the configuration.
+"""
+
+from qm.qua import *
+from qm import QuantumMachinesManager
+from configuration_with_octave import *
+from qualang_tools.results import progress_counter, wait_until_job_is_paused
+from qualang_tools.plot import interrupt_on_close
+from qualang_tools.loops import from_array
+import matplotlib.pyplot as plt
+from scipy.signal import detrend
+
+
+###################
+# The QUA program #
+###################
+
+n_avg = 100  # The number of averages
+# The intermediate frequency sweep parameters
+f_min = 21 * u.MHz
+f_max = 271 * u.MHz
+df = 500 * u.kHz
+IFs = np.arange(f_min, f_max + 0.1, df)  # The intermediate frequency vector (+ 0.1 to add f_max to IFs)
+# This is to make sure that the center IF is the one used in the config for the correction parameters to be updated.
+config["elements"]["resonator"]["intermediate_frequency"] = IFs[len(IFs) // 2]
+
+# The LO frequency sweep parameters
+f_min_lo = 4.0e9
+f_max_lo = 5.0e9
+df_lo = f_max - f_min
+LOs = np.arange(f_min_lo, f_max_lo + 0.1, df_lo)
+frequency = np.array(np.concatenate([IFs + LOs[i] for i in range(len(LOs))]))
+
+with program() as resonator_spec:
+    n = declare(int)  # QUA variable for the averaging loop
+    i = declare(int)  # QUA variable for the LO frequency sweep
+    f = declare(int)  # QUA variable for the resonator IF
+    I = declare(fixed)  # QUA variable for the measured 'I' quadrature
+    Q = declare(fixed)  # QUA variable for the measured 'Q' quadrature
+    I_st = declare_stream()  # Stream for the 'I' quadrature
+    Q_st = declare_stream()  # Stream for the 'Q' quadrature
+    n_st = declare_stream()  # Stream for the averaging iteration 'n'
+
+    with for_(i, 0, i < len(LOs) + 1, i + 1):
+        pause()  # This waits until it is resumed from python
+        with for_(n, 0, n < n_avg, n + 1):
+            with for_(*from_array(f, IFs)):
+                # Update the frequency of the digital oscillator linked to the qubit element
+                update_frequency("resonator", f)
+                # Measure the state of the resonator
+                measure(
+                    "readout",
+                    "resonator",
+                    None,
+                    dual_demod.full("cos", "out1", "sin", "out2", I),
+                    dual_demod.full("minus_sin", "out1", "cos", "out2", Q),
```

Comment:
This is written for qua version <1.2
Should we update it to the new syntax? @TheoLaudatQM

Reply:
Good point, but actually 1.2 and QOP2.4.x are not officially out yet so I am not sure
I am in favor of updating to the newest syntax and wait until it is officially released before pushing it, what do you think?

Reply:
We live it like this